ocsp/responder/multi_source_test.go

```diff
@@ -13,38 +13,122 @@ import (
 	"golang.org/x/crypto/ocsp"
 )
 
-type succeedSource struct {
-	resp *Response
+type ok struct{}
+
+const expectedFreshness = 60 * time.Hour
+
+func (src ok) Response(context.Context, *ocsp.Request) (*Response, error) {
+	return &Response{
+		Response: &ocsp.Response{
+			Status:     ocsp.Good,
+			ThisUpdate: time.Now().Add(-10 * time.Hour),
+		},
+		Raw: nil,
+	}, nil
 }
 
-func (src *succeedSource) Response(context.Context, *ocsp.Request) (*Response, error) {
-	if src.resp != nil {
-		return src.resp, nil
-	}
-	// We can't just return nil, as the multiSource checks the Statuses from each
-	// Source to ensure they agree.
-	return &Response{&ocsp.Response{Status: ocsp.Good}, []byte{}}, nil
+type revoked struct{}
+
+func (src revoked) Response(context.Context, *ocsp.Request) (*Response, error) {
+	return &Response{
+		Response: &ocsp.Response{
+			Status:     ocsp.Revoked,
+			ThisUpdate: time.Now().Add(-10 * time.Hour),
+		},
+		Raw: nil,
+	}, nil
+}
+
+type stale struct{}
+
+func (src stale) Response(context.Context, *ocsp.Request) (*Response, error) {
+	return &Response{
+		Response: &ocsp.Response{
+			Status:     ocsp.Good,
+			ThisUpdate: time.Now().Add(-70 * time.Hour),
+		},
+		Raw: nil,
+	}, nil
 }
 
-type failSource struct{}
+type fail struct{}
 
-func (src *failSource) Response(context.Context, *ocsp.Request) (*Response, error) {
+func (src fail) Response(context.Context, *ocsp.Request) (*Response, error) {
 	return nil, errors.New("failure")
 }
 
-// timeoutSource is a Source that will not return until its chan is closed.
-type timeoutSource struct {
+// timeout is a Source that will not return until its chan is closed.
+type timeout struct {
 	ch <-chan struct{}
 }
 
-func (src *timeoutSource) Response(context.Context, *ocsp.Request) (*Response, error) {
+func (src timeout) Response(context.Context, *ocsp.Request) (*Response, error) {
 	<-src.ch
 	return nil, errors.New("failure")
 }
 
+func TestMultiSource(t *testing.T) {
+	type testCase struct {
+		primary        Source
+		secondary      Source
+		expectedError  bool
+		expectedStatus int // only checked if err is nil
```

Comment:
nit: only checked if expectedError is false

Reply:
"Only checked if err is nil" is correct - the `if err != nil` clause ensures we bail out early in the other case. Or are you saying the tests would be easier to understand if they had the property you propose?

Reply:
The tests already do have this property: if `expectedError` is true and we got an error, we bail out on line 117; if `expectedError` is true and we didn't get an error, we bail out on line 120.

But yeah I'm mostly just saying that I think it's less confusing to say "we only care about the expected status if we don't expect an error" than "we only care about the expected status if we didn't _get_ an error". Means that the semantics of these two fields are only dependent on each other, not dependent on test behavior.